cub/cub/detail/rfa.cuh

```diff
@@ -0,0 +1,950 @@
+  }
+
+  /// Return carry vector value ref
+  _CCCL_DEVICE _CCCL_FORCEINLINE ftype& carry(int i)
+  {
+    const auto& c = *this;
+    return const_cast<ftype&>(c.carry(i));
+  }
+
+#ifdef DISABLE_ZERO
+  _CCCL_DEVICE _CCCL_FORCEINLINE static constexpr bool is_zero_v(const ftype)
+  {
+    return false;
+  }
+#else
+  _CCCL_DEVICE _CCCL_FORCEINLINE static constexpr bool is_zero_v(const ftype x)
+  {
+    return x == 0.0;
+  }
+#endif
+
+#ifdef DISABLE_NANINF
+  _CCCL_DEVICE _CCCL_FORCEINLINE static constexpr int is_nan_inf_v(const ftype)
+  {
+    return false;
+  }
+#else
+  _CCCL_DEVICE _CCCL_FORCEINLINE static constexpr int is_nan_inf_v(const ftype x)
+  {
+    const auto bits = get_bits(x);
+    return (bits & ((2ull * MAX_EXP - 1) << (MANT_DIG - 1))) == ((2ull * MAX_EXP - 1) << (MANT_DIG - 1));
+  }
+#endif
+
+  _CCCL_DEVICE _CCCL_FORCEINLINE static int exp_val(const ftype x)
+  {
+    const auto bits = get_bits(x);
+    return (bits >> (MANT_DIG - 1)) & (2 * MAX_EXP - 1);
+  }
+
+  /// Get index of float-point precision
+  /// The index of a non-binned type is the smallest index a binned type would
+  /// need to have to sum it reproducibly. Higher indices correspond to smaller
+  /// bins.
+  _CCCL_DEVICE _CCCL_FORCEINLINE static int binned_dindex(const ftype x)
+  {
+    int exp = exp_val(x);
+    if (exp == 0)
+    {
+      if (x == 0.0)
+      {
+        return MAXINDEX;
+      }
+      else
+      {
+        (void) ::cuda::std::frexpf(x, &exp);
+        // use +MAXINDEX to optimize the host constexpr variable `MAXINDEX` to be used in the device function
+        return (::cuda::std::min)((MAX_EXP - exp) / BIN_WIDTH, +MAXINDEX);
+      }
+    }
+    return ((MAX_EXP + EXP_BIAS) - exp) / BIN_WIDTH;
+  }
+
+  /// Get index of manually specified binned double precision
+  /// The index of a binned type is the bin that it corresponds to. Higher
+  /// indices correspond to smaller bins.
+  _CCCL_DEVICE _CCCL_FORCEINLINE int binned_index() const
+  {
+    return ((MAX_EXP + MANT_DIG - BIN_WIDTH + 1 + EXP_BIAS) - exp_val(primary(0))) / BIN_WIDTH;
+  }
+
+  /// Check if index of manually specified binned floating-point is 0
+  /// A quick check to determine if the index is 0
+  _CCCL_DEVICE _CCCL_FORCEINLINE bool binned_index0() const
+  {
+    return exp_val(primary(0)) == MAX_EXP + EXP_BIAS;
+  }
+
+  //! Update manually specified binned fp with a scalar (X -> Y)
+  //!
+  //! This method updates the binned fp to an index suitable for adding numbers
+  //! with absolute value less than @p max_abs_val
+  //!
+  //!@param incpriY stride within Y's primary vector (use every incpriY'th element)
+  //!@param inccarY stride within Y's carry vector (use every inccarY'th element)
+  _CCCL_DEVICE void binned_dmdupdate(const ftype max_abs_val, const int incpriY, const int inccarY)
+  {
+    if (is_nan_inf_v(primary(0)))
+    {
+      return;
+    }
+
+    int X_index = binned_dindex(max_abs_val);
+    if (is_zero_v(primary(0)))
+    {
```

Comment:
Nitpick: This if is the last statement in the function. We could simplify the rest of the function by adding a `return;` and removing the `else`